Adding Cluster API doc entry (#1291)
giantswarm/docs: src/content/general/architecture/capi/index.md

---
linkTitle: Cluster-API-Architecture
title: The Giant Swarm Platform Architecture

> description: Architecture overview explaining how our Platform is built it.

Author: ```suggestion
description: Architecture overview explaining how our Platform is built and what services do we offer.
```

## Cluster Architecture

As explained previously, both the management cluster and the workload cluster have the same cluster structure and configuration. In Giant Swarm we rely on [Cluster API](https://cluster-api.sigs.k8s.io/) to bootstrap and configure the cluster infrastructure and set up all the components needed for a cluster to function.

[Cluster Architecture Image](./CAPI_architecture.png)!

By default, the machines are split into three different failure domains or zones to ensure the availability of the API and workloads running on the there. In our setup, three control plane machines hold the Kubernetes API and the other controllers, and a variable number of worker machines contain the regular services.

There is a machine created as a bastion that helps us with the operations. It is the single entry point to the running infrastructure so that way all the cluster machines can live in a private network and expose the services running on them via explicit configuration.

{{< tabs >}}
{{< tab id="flags-aws" title="AWS">}}

> 

Author: @paurosello can you help me to fill this part, how CAPA connect to the Internet?

{{< /tab >}}
{{< tab id="flags-capz" title="Azure">}}

{{< /tab >}}
{{< tab id="flags-gcp" title="GCP">}}

> 

Author: @fiunchinho  can you help me to fill this part, how CAPG networking works in general?

There are three types of Giant Swarm components running on the management clusters: the operators, the admission controllers and the operational services. The first one, operators, extend the API allowing our customers to manage new entities (like a Cluster or an App) as if they were built-in resources. The second one, admission controllers, validate and default the resources to create a better user experience. And the last one, the operational services, monitoring or security, enable our staff to ensure all the workload clusters and applications are up and running securely and seamlessly.

##### Operators (Custom Resource plus controller)

The most common pattern to extend Kubernetes is called [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/). It allows to define a new [Kubernetes resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) (called Custom Resource) and apply functionality to it. 

In Giant Swarm we leverage the operator pattern to extend the management cluster and provide a great user experience to our customers.

*Organization operator*

> This operator is in charge of reconciling `Organization` custom resources. The functionality of the operator is pretty straightforward. It takes creates and delete the organization namespace when the given resource is created.

Reviewer: ```suggestion
This operator is in charge of reconciling `Organization` custom resources. The functionality of the operator is pretty straightforward. It creates and deletes the organization namespace when the given resource is created.
```

To learn more about organizations please read [the related documentation](https://docs.giantswarm.io/general/organizations/).

*RBAC operator*

We have built an RBAC operator with the goal of maintaining up to date permissions between the different organizations and users on the management cluster so [you can isolate your teams and ensure access level in a granular way](https://docs.giantswarm.io/ui-api/management-api/authorization/).

*DNS operator*

We built an operator to manage the DNS entries for all the endpoints we need to expose to our customers. All our management clusters have a base domain used to allocate the API, OIDC endpoint or UI. This operator listens to cluster resource and provides the right records.

*Deletion Blocker Operator*

> As a Kubernetes user, you have most likely experienced a case in which a resource block deletion. This typically happens because the back-end resource is not there or it is in `failed` state. This operator ensure all leftovers are cleaned up after a cluster is marked for deletion. 

Reviewer: ```suggestion
Kubernetes ensures resources are deleted in the right order by using `finalizers`, a way to create dependencies between resources. In specific cases, some CRs don't leverage finalizers, which can stop the deletion process when they are deleted too early and leave the cluster in an inconsistent state. Deletion Blocker Operator ensure that specific CRs aren't deleted until specific conditions are met.
```
This was the description of `cleanup-operator`.

*App operator*

The [App Platform](https://docs.giantswarm.io/app-platform/) is a system we have built to deliver Cloud Native Apps in multi cluster fashion. App operator is the main code running in management cluster to manage App custom resources and make sure the applications and configuration are delivered correctly across the different workload clusters.

*Cluster API operators*

In [Cluster API](https://cluster-api.sigs.k8s.io/) there is a set of generic operators that all Cluster API providers run to address the common bootstrap and management actions on a Kubernetes cluster lifecycle which include:

- `Kubeadm Bootstrap`controller which is in charge of providing a cloud init configuration to turn a machine into a Kubernetes node.
- `Kubeadm Control Plane` controller that manages the lifecycle of the control plane nodes and provides access to the API.
- `CAPI manager` controller that manages cluster and machine resources.

*Cluster OpenStack controller*

> As part of the generic controllers we have this special one. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster. This controller provisions, updates and deletes all resources on the provider API. It also enriches Kubernetes node info with infrastructure relevant information. This is helpful for application administrators to implement affinity/ anti-affinity rules which reflect underlying availability zone metadata. It is mandatory for the Cluster API provisioning workflow to get known infrastructure specific machine metadata.

Author: this is a an upstream controller

> As part of the generic controllers we have this special one. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster. This controller provisions, updates and deletes all resources on the provider API. It also enriches Kubernetes node info with infrastructure relevant information. This is helpful for application administrators to implement affinity/ anti-affinity rules which reflect underlying availability zone metadata. It is mandatory for the Cluster API provisioning workflow to get known infrastructure specific machine metadata.

Author: I made this generic for all providers

> As part of the generic controllers we have this special one. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster. This controller provisions, updates and deletes all resources on the provider API. It also enriches Kubernetes node info with infrastructure relevant information. This is helpful for application administrators to implement affinity/ anti-affinity rules which reflect underlying availability zone metadata. It is mandatory for the Cluster API provisioning workflow to get known infrastructure specific machine metadata.

Author: in future we can extend it and be more specific for each provider individually (tabs)

> In Cluster API there is a set of common providers defined above, but to manage the real infrastructure we need to run an specific controller that knows about the provider API. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster. This controller provisions, updates and deletes all resources through API. Every provider has its own lingo and methods to create a machine or provision a network. The operator reconciles the `Infrastructure` custom resources of the specific provider parsing the desired state into API requests and ensuring the provider resources are always in sync with the resource definitions stored in the management cluster.

Reviewer: ```suggestion
In Cluster API there is a set of common providers defined above, but to manage the real infrastructure we need to run a specific controller that knows about the cloud provider's API. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster (i.e. VMs, Load Balancers...). This controller provisions, updates and deletes all resources through API. Every provider has its own lingo and methods to create a machine or provision a network. The operator reconciles the `Infrastructure` custom resources of the specific provider parsing the desired state into API requests and ensuring the provider resources are always in sync with the resource definitions stored in the management cluster.
```

*Cluster apps operator*

> This simple operator takes care of providing the basic configuration for our managed apps. Assets like the certificate authority or the domain base are provided by it via a configmap so the different apps can rely on those values to configure themselves.

Reviewer 2: We don't have rules for these specific cases, just guidelines (more for blog posts) - https://docs.google.com/document/d/1TKzd70koVmrJyXx0JOe15tK0-mIiGYkP3gz_2BxM844/edit
I suggest choosing one way and being consistent throughout the document.

> This simple operator takes care of providing the basic configuration for our managed apps. Assets like the certificate authority or the domain base are provided by it via a configmap so the different apps can rely on those values to configure themselves.

Author: wow that's document is gold

##### Admission controllers

The [Admission Controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) is other feature Kubernetes offers to extend its functionality. This time the idea is to intercept the API requests for validating or mutating the content. The main reason is to avoid users submit wrong data or default some of the properties for custom resources. 

*App Admission Controller*

> This admission controller gives the ability to our App Platform to ensure an App is created properly, with all required files and points to an existing app in the Catalog. At the same time allow our customers to only passed the minimum spec and default the rest automatically for them.

Author: yeah that is a good idea

> This admission controller gives the ability to our App Platform to ensure an App is created properly, with all required files and points to an existing App in the Catalog. At the same time allow our customers to only passed the minimum spec and default the rest automatically.

Reviewer: ```suggestion
This admission controller gives the ability to our App Platform to ensure an App is created properly, with all required files and points to an existing App in the Catalog. At the same time, it allows our customers to pass the bare minimum required spec and will default the rest automatically.
```

*Cluster API admission controllers*

Each CAPI controller has its own admission to enhance the cluster management experience. Basically it perform same action of validation and mutation. 

*Management admission controller*

As we described before we have organization to organize clusters and apps, but we need an admission to make sure some rules are respected. For example, you cannot delete an organization if this one has cluster running on it. This controller make sure the request respect those rules.

*Kyverno admission controller*

> As you my hear before [Pod Security Policies are deprecated](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/) and from Kubernetes `1.25` they will not be valid anymore. We have relied on them since the beginning and we harden the cluster to make sure containers run with the least privileges as possible. For that reason we have started to implement the same policies in [Kyverno](https://kyverno.io/). 

Reviewer 2: Do we really need the whole history of PSP deprecation for a CAPI doc? Have any of our users used PSP for CAPI? I know the answer for production is no.
So is this even relevant, or can it be phrased differently and just discuss how we solve role based acces?

> As you my hear before [Pod Security Policies are deprecated](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/) and from Kubernetes `1.25` they will not be valid anymore. We have relied on them since the beginning and we harden the cluster to make sure containers run with the least privileges as possible. For that reason we have started to implement the same policies in [Kyverno](https://kyverno.io/). 

Author: So we are transitioning from old PSP to Kyverno, IMO till we finally moved to Kyverno I would explain it

> [Pod Security Policies are deprecated](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/) and from Kubernetes `1.25` they will not be valid anymore. In the past relied on this functionality to harden the cluster and ensure containers run with the least privileges as possible. As a replacement we implemented the same policies in [Kyverno](https://kyverno.io/). 

Reviewer: ```suggestion
[Pod Security Policies are deprecated](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/) and from Kubernetes `1.25` they will not be valid anymore. In the past we relied on this functionality to harden the cluster and ensure containers run with the least privileges required. As a replacement we implemented equivalent policies in [Kyverno](https://kyverno.io/). 
```

##### Operational services

> Aside from operators and admission controllers we also deploy a set of tooling that ensure we have a nice delivery system, good hardening and decent observability.  

Author: ```suggestion
Aside from operators and admission controllers we also deploy a set of tooling that ensure we have a nice delivery system, good hardening and clear observability.  
```

> Aside from operators and admission controllers we also deploy a set of tooling that ensure we have a nice delivery system, good hardening and decent observability.  

Reviewer 2: IMO decent is a mediocre word. I'd use clear as the adjective for visibility.

*Monitoring Features of the Platform*

Since we provide a **managed** Kubernetes platform, Giant Swarm has to be aware of state and unexpected events regarding the platform. For that reason our management clusters run a [monitoring stack](https://www.giantswarm.io/blog/monitoring-on-demand-kubernetes-clusters-with-prometheus) to watch all workload clusters and ensure all managed components are healthy. In each workload cluster there are several [exporters](https://prometheus.io/docs/instrumenting/exporters/) that gather and forward the metrics for each component.

> Our on-call engineers will be paged in case anything happens to the cluster or its base components and they will respond to the incident based on the run-books we have created based on years of operating Cloud Native systems. In case there is an improvement to be made, a post mortem is created and a solution will be implemented before long. Any patch or fix added to the platform will be released to all customers.

Reviewer 2: ```suggestion
Our engineers are that are on-call are paged in case anything happens to the cluster or its base components. They respond to the incident based on our run-books. The ones we have written (and continue to update) over years operating Cloud Native systems. In case there is an improvement to be made, a post mortem is created and a solution will be implemented before long. Any patch or fix added to the platform is released to all customers.
```

> Our engineers are that are on-call are paged in case anything happens to the cluster or its base components. They respond to the incident based on our run-books. The ones we have written (and continue to update) over years operating Cloud Native systems. In case there is an improvement to be made, a post mortem is created and a solution will be implemented before long. Any patch or fix added to the platform is released to all customers.

Reviewer: ```suggestion
Our on-call engineers are paged in case anything happens to the cluster or its base components. They respond to the incident based on our run-books, the ones we have written (and continue to update) over years operating Cloud Native systems. In case there is an improvement to be made, a post mortem is created and a solution will be implemented before long. Any patch or fix added to the platform is released to all customers.
```

*CI/CD Features of the Platform*

> Since the appearance of [GitOps](https://www.giantswarm.io/blog/what-is-gitops] we have been enthusiastic about it. It provides many benefits and rely in the same principles we were already advocating for. In Giant Swarm, we use [Flux](https://www.giantswarm.io/blog/gitops-with-flux-giant-swarm) to control the configuration and definition of our infrastructure and software on top of it. 

Reviewer 2: ```suggestion
Since the appearance of [GitOps](https://www.giantswarm.io/blog/what-is-gitops] we have been enthusiastic about it. It provides many benefits while relying on the same principles we were already advocating for. In Giant Swarm, we use [Flux](https://www.giantswarm.io/blog/gitops-with-flux-giant-swarm) to control the configuration and definition of infrastructure and the software on top of it. 
```

> In our setup we have two Flux instances running, one managing the ...

Author: @mproffitt can you help to fill this part?

---

> linkTitle: Cluster-API-Architecture

Reviewer 3: ```suggestion
linkTitle: Cluster API architecture
```

> title: The Giant Swarm Platform Architecture

Reviewer 3: This title should differentiate, as we have some other architecture docs/versions, too.

Also, please use sentence case.

```suggestion
title: The Giant Swarm platform architecture using Cluster API
```

description: Architecture overview explaining how our Platform is built and what services do we offer.
weight: 30
menu:
  main:
    parent: general-architecture
last_review_date: 2022-09-12
user_questions:

>   - Do you run a Developer Platform?

Reviewer 3: This user question is very general. I suggest something specific to Cluster API.

> aliases:
>   - /basics/platform-architecture/
> owner:

Reviewer 3: ```suggestion
owner:
```

No need for aliases here. aliases are created in case a page is moved, to have a redirect from the old URL to the new one.

  - https://github.com/orgs/giantswarm/teams/team-rocket
---

> The Giant Swarm Platform consists of various systems. They can be categorized into three areas: infrastructure, applications, and operations.

Reviewer 3: Please use sentence case everywhere, except for brand names. These are brand names:

- Giant Swarm
- Kubernetes
- Cluster API

These are not:

- cloud native
- platform

Giant Swarm's architecture is split into two logical parts. One encompasses the management cluster and all the components running there. The second part  refers to the workload clusters that are created dynamically by the users to run their business workloads. In principle the management cluster and workload cluster(s) are analogous in terms of infrastructure and configuration. The difference comes with the additional layers we deployed on top of the management cluster that helps manage your users and permissions, workload clusters or the applications running on the workload clusters.

## Cluster Architecture

As explained previously, both the management cluster and the workload cluster(s) have the same structure and configuration. In Giant Swarm we rely on [Cluster API](https://cluster-api.sigs.k8s.io/) to bootstrap and configure the cluster infrastructure and set up all the components needed for a cluster to function.

> [Cluster Architecture Image](./CAPI_architecture.png)!

Reviewer 3: ```suggestion
![Cluster Architecture Image](CAPI_architecture.png)
```

{{< /tab >}}
{{< tab id="flags-capz" title="Azure">}}

{{< /tab >}}
{{< tab id="flags-gcp" title="GCP">}}

The setup uses the default network if none has been created upfront for the purpose. Every Google project always come with a default network. The network allows the machines to pull images or route requests from containers to the Internet. The machines are running in private subnets configured to use a Cloud NAT gateway to access the outside network. At the same time, Google uses Global Load Balancer to route external requests to the containers.

{{< /tab >}}
{{< /tabs >}}

Internally, Cluster API (CAPI) uses kubeadm to configure all the machines according to current standards. It uses a template defined as yaml, part of CAPI, where we have hardened the different parameters for the API and other controllers running in the master machine.

## Giant Swarm Platform

The Giant Swarm Platform is based on the API of the management cluster. The main reason is that Kubernetes has become the de-facto standard for managing infrastructure in a modern way. Its extensibility makes it easy to transform a cluster in a Platform with a secret sauce that makes the experience great.

> Our platform let's customers manage (workload) clusters and applications in a Cloud Native fashion. Following is an explanation of the bootstrapping process of a cluster, how it is managed throughout its lifecycle and which cluster components run based on its role (i.e. workload cluster or management cluster).

Reviewer 3: ```suggestion
Our platform lets customers manage (workload) clusters and applications in a Cloud Native fashion. Following is an explanation of the bootstrapping process of a cluster, how it is managed throughout its lifecycle and which cluster components run based on its role (i.e. workload cluster or management cluster).
```

2. Add the installation details into our config management system (Github repo).
3. Create a bootstrap cluster using [kind](https://kind.sigs.k8s.io/).
4. Install the Cluster API (CAPI) controllers on it.
5. Create the management CAPI resources on the bootstrapping cluster to trigger the provision of the real infrastructure on the provider. It creates the actual Kubernetes cluster in the customer infrastructure.
6. Install the Cluster API controllers on the real management cluster and move the resources to it.
7. Deploy our App platform on top of the management cluster.
8. Deploy our monitoring system on top of the management cluster.
9. Deploy all additional operators that enhances the API (organizations, RBAC, etc).
10. Configure and harden the cluster.
11. Run and test the management cluster functionality.

After the management cluster is ready we deliver all the details to customers to give them the access to the Management API. From this point we use the same mechanisms to control the management cluster lifecycle as we use for workload clusters.

Cluster API offers a set of custom resources that define all the details of a cluster infrastructure and its configuration. 

> [Cluster API Resources](./CAPI_resources.png)!

Reviewer 3: ```suggestion
![Cluster API resources](CAPI_resources.png)
```

Usually we have generic resources that define common configuration of the clusters and its components, and some infrastructure specific resources which will be tied to the provider we select. 

> We have created a [kubectl plugin](https://docs.giantswarm.io/ui-api/kubectl-gs/template-cluster/) to help you template all the needed resources.

Reviewer 3: Please use the `relref` syntax for internal links.

```suggestion
We have created a [kubectl plugin]({{< relref "/ui-api/kubectl-gs/template-cluster" >}}) to help you template all the needed resources.
```

##### Operators (Custom Resource plus controller)

The most common pattern to extend Kubernetes is called [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/). It allows to define a new [Kubernetes resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) (called Custom Resource) and apply functionality to it. 

In Giant Swarm we leverage the operator pattern to extend the management cluster and provide a great user experience to our customers.

*Organization operator*

This operator is in charge of reconciling `Organization` custom resources. The functionality of the operator is pretty straightforward. It creates and deletes the organization namespace when the given resource is created.

> To learn more about organizations please read [the related documentation](https://docs.giantswarm.io/general/organizations/).

Reviewer 3: Please use relref

*RBAC operator*

> We have built an RBAC operator with the goal of maintaining up to date permissions between the different organizations and users on the management cluster so [you can isolate your teams and ensure access level in a granular way](https://docs.giantswarm.io/ui-api/management-api/authorization/).

Reviewer 3: relref

*DNS operator*

We built an operator to manage the DNS entries for all the endpoints we need to expose to our customers. All our management clusters have a base domain used to allocate the API, OIDC endpoint or UI. This operator listens to cluster resource and provides the right records.

*Deletion Blocker Operator*

Kubernetes ensures resources are deleted in the right order by using `finalizers`, a way to create dependencies between resources. In specific cases, some CRs don't leverage finalizers, which can stop the deletion process when they are deleted too early and leave the cluster in an inconsistent state. Deletion Blocker Operator ensure that specific CRs aren't deleted until specific conditions are met.

*App operator*

> The [App Platform](https://docs.giantswarm.io/app-platform/) is a system we built to deliver Cloud Native Apps in a multi cluster fashion. App operator is the main code running in management clusters to manage App custom resources and make sure the applications and configuration are delivered correctly across the different workload clusters.

Reviewer 3: relref

*Cluster API operators*

In [Cluster API](https://cluster-api.sigs.k8s.io/) there is a set of generic operators that all Cluster API providers run to address the common bootstrap and management actions on a Kubernetes cluster lifecycle which include:

> - `Kubeadm Bootstrap`controller which is in charge of providing a cloud init configuration to turn a machine into a Kubernetes node.

Reviewer 3: ```suggestion
- `Kubeadm Bootstrap` controller which is in charge of providing a cloud init configuration to turn a machine into a Kubernetes node.
```

Reviewer 3: In general, I would question the use of backticks here. We use them to indicate that something is "code", to show that exact spelling is important.

If you just want to emphasise, you can use `**` to make the name bold.

- `Kubeadm Control Plane` controller that manages the lifecycle of the control plane nodes and provides access to the API.
- `CAPI manager` controller that manages cluster and machine resources.

*Cluster Provider operator*

> In Cluster API there is a set of common providers defined above, but to manage the real infrastructure we need to run a specific controller that knows about the cloud provider's API. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster (i.e. VMs, Load Balancers...). This controller provisions, updates and deletes all resources through API. Every provider has its own lingo and methods to create a machine or provision a network. The operator reconciles the `Infrastructure` custom resources of the specific provider parsing the desired state into API requests and ensuring the provider resources are always in sync with the resource definitions stored in the management cluster.

Reviewer 3: ```suggestion
In Cluster API there is a set of common providers defined above, but to manage the real infrastructure we need to run a specific controller that knows about the cloud provider's API. It acts as a bridge and reconciles the provider specific configuration to create the necessary infrastructure for a Kubernetes cluster (i.e. VMs, Load Balancers...). This controller provisions, updates and deletes all resources through API. Every provider has its own lingo and methods to create a machine or provision a network. The operator reconciles the infrastructure resources of the specific provider parsing the desired state into API requests and ensuring the provider resources are always in sync with the resource definitions stored in the management cluster.
```

Since there is no resource named "Infrastructure".

---
linkTitle: Cluster API architecture
title: The Giant Swarm platform architecture using Cluster API
description: Architecture overview explaining how our platform is built and what services do we offer.
weight: 30
menu:
  main:
    parent: general-architecture
last_review_date: 2022-09-12
user_questions:
  - How do you run a developer platform based on Cluster API?

>   - How Giant Swarm has built a platform to built platforms with Cluster API?

Reviewer 3: ```suggestion
  - How has Giant Swarm built a platform to build platforms with Cluster API?
```
